File: search_dfs_bfs/lv2_소수찾기.py
#https://programmers.co.kr/learn/courses/30/lessons/42839
from itertools import permutations

def is_prime(n):
    if n == 0 or n == 1:
        return False
    else:
        for i in range(2, n):
            if n % i == 0:
                return False
        return True

def solution(numbers):
    result = set()
    for i in range(1, len(numbers) + 1):
        p = permutations(numbers, i)
        result |= {int("".join(n)) for n in p}
    return sum([1 for n in list(result) if is_prime(n)])

File: search_dfs_bfs/lv2_소수찾기.java
// https://programmers.co.kr/learn/courses/30/lessons/42839

import java.util.HashSet;


class Solution {
    public int solution(String numbers) {
        int answer = 0;
        HashSet<Integer> set = new HashSet<>();
        permutation("", numbers, set);
        while (set.iterator().hasNext()) {
            int now = set.iterator().next();
            if (isPrime(now)) {
                answer ++;
            }
            set.remove(now);
        }
        return answer;
    }

    public boolean isPrime(int num) {
        if (num <= 1) return false;
        for (int i = 2; i <= (int) Math.sqrt(num); i++) {
            if (num % i == 0) return false;
        }
        return true;
    }

    public void permutation(String begin, String end, HashSet<Integer> result) {
        if (!begin.equals("")) result.add(Integer.valueOf(begin));
        for (int i = 0; i < end.length(); i++) {
            permutation(begin + end.charAt(i), end.substring(0, i) + end.substring(i+1, end.length()), result);
        }
    }
}
